### PR TITLE
Sourcecode for metric onboarding check

### PR DESCRIPTION
**Pull request by tusharbahl**

### CODE REVIEW

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    @Override
+    public Integer onboard(Metric metric) {
+        List<Tag> tagList;
+        Metric metricEntry;
+        Tag tagEntry;
+        Integer metricID = 0;
+
+
+        for (Iterator<Map.Entry<String, Object>> tagIterator = metric.getTags().entrySet().iterator(); tagIterator.hasNext(); ) {
+            Map.Entry<String, Object> entry = tagIterator.next();
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java]: on second thought I think it's better if we keep only one endpoint /onboard in model service. if the metric is not onboard we onboard it and returns the metric id. Else we return ItemExistsException.

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        Metric m = metricRepository.findByHash(hash);
+        System.out.println("Metric is: "+m);
+        if (m != null) {
+            System.out.println("Metric Exist");
+            throw new ItemExistsException(m);
+        }
+        List<Tag> tagList;
+        Metric metricEntry;
+        Tag tagEntry;
+        Long metricID = Long.valueOf(0);
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java]: Move all the variables to the top of the function and initialize list and objects here.

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/repo/TagRepository.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * Spring Data repository for tags.
+ *
+ * @author tbahl
+ */
+@RepositoryRestResource(excerptProjection = InlineType.class)
+public interface TagRepository extends PagingAndSortingRepository<Tag, Long> {
+
+
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/repo/TagRepository.java]: Add some information about what this method does

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    private MetricRepository metricRepository;
+
+    @Autowired
+    private MetricTagMappingRepository metricTagMappingRepository;
+
+    @Override
+    public Metric onboard(Metric metric) {
+        List<Tag> tagList = new ArrayList<>();
+        Metric metricEntry = null;
+        Tag tagEntry;
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java]: initialize metric and tag objects

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        String hash = metric.getHash();
+        Metric m = metricRepository.findByHash(hash);
+        if (m != null) {
+            throw new ItemExistsException(m);
+        }
+        for (Iterator<Map.Entry<String, Object>> tagIterator = metric.getTags().entrySet().iterator(); tagIterator.hasNext(); ) {
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java]: something like below would be easy to read.
Tags tags =  metric.getTags();
Iterator iterator = .entrySet().iterator()
and then use these values in for loop

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            Map.Entry<String, Object> entry = tagIterator.next();
+
+            metricRepository.save(metric);
+            metricEntry = metricRepository.findByHash(metric.getHash());
+            tagList = tagRepository.findByUkeyContainingAndUvalueContaining(entry.getKey(), (String) entry.getValue());
+
+            if (tagList.size() == 0) {
+                tagEntry = tagRepository.save(new Tag(entry.getKey(), (String) entry.getValue()));
+                metricTagMappingRepository.save(new MetricTagMapping(metricEntry, tagEntry));
+            } else { metricTagMappingRepository.save(new MetricTagMapping(metricEntry, tagList.get(0))); }
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java]: Run formatter here

Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java]: See if you can skip 1 else here. Ideally, if the tag doesn't exist then we need to add an entry in the tag table. We need to add mapping irrespective of the fact whether the tag exists or not.

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    private MetricTagMappingRepository metricTagMappingRepository;
+
+    @Override
+    public Metric onboard(Metric metric) {
+        Metric m = metricRepository.findByHash(metric.getHash());
+        if (m != null) {
+            throw new ItemExistsException(m);
+        }
+
+        List<Tag> tagList = new ArrayList<>();
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java]: @tusharbahl rename this to tags.

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        Metric metricEntry = new Metric();
+        Tag tagEntry = new Tag();
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java]: rename this to tag

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    public Metric onboard(Metric metric) {
+        Metric m = metricRepository.findByHash(metric.getHash());
+        if (m != null) {
+            throw new ItemExistsException(m);
+        }
+
+        List<Tag> tagList = new ArrayList<>();
+        Metric metricEntry = new Metric();
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java]: rename this to metric

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        Iterator<Map.Entry<String, Object>> tagiterator = tag.entrySet().iterator();
+
+        for(Iterator<Map.Entry<String, Object>> iterator = tagiterator;  iterator.hasNext();) {
+            Map.Entry<String, Object> entry = iterator.next();
+            metricRepository.save(metric);
+            metricEntry = metricRepository.findByHash(metric.getHash());
+            tagList = tagRepository.findByUkeyContainingAndUvalueContaining(entry.getKey(), (String) entry.getValue());
+
+            if (tagList.size() == 0) {
+                tagEntry = tagRepository.save(new Tag(entry.getKey(), (String) entry.getValue()));
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java]: see if we can call metricTagMappingRepository only once here.

File: modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/web/OnboardController.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import com.expedia.adaptivealerting.modelservice.entity.Metric;
+import com.expedia.adaptivealerting.modelservice.service.OnboardService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Controller for OnboardService.
```
Comment (karanshah03) [inline modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/web/OnboardController.java]: add bit more info about this controller.

**Comment by tusharbahl** [comment]
Closing this PR, opened a new one #296

**Pull request closed without merging**